### ISSUE TITLE
netstandard2.1 incompatibility with 5.0.0 packages is a maintenance issue

### ISSUE THREAD

**Issue by stefanloerwald** [dotnet/aspnetcore#27717]
This is a usability issue / maintenance issue. I personally maintain a few blazor component libraries and so far have targetted `netstandard2.1`. Now the nuget packages (e.g. `Microsoft.AspNetCore.Components.Web`) have been updated to version 5.0.0, corresponding to `net5.0`. I would love to jump on the train to .NET 5, but my options here are all problematic:

### 1. Staying on `netstandard2.1`:

The nuget package manager and other tools (e.g. dependabot) will suggest the update to package version 5.0.0, even though that will fail due to incompatibility. There are now updates available to 3.1.X (namely currently 3.1.10), but they are less discoverable, because 5.0.0 is considered more recent by the package management tools.

### 2. Dual-targetting `netstandard2.1` and `net5.0`:

While it's possible to do this, the csproj becomes quite verbose:
```xml
<PropertyGroup>
	<TargetFrameworks>netstandard2.1;net5.0</TargetFrameworks>
        <RazorLangVersion Condition= " '$(TargetFramework)' != 'net5.0' ">3.0</RazorLangVersion>
</PropertyGroup>
<ItemGroup>
	<PackageReference Include="Microsoft.AspNetCore.Components" Version="3.1.9" Condition= " '$(TargetFramework)' != 'net5.0' " />
	<PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="3.1.9" Condition= " '$(TargetFramework)' != 'net5.0' " />
	<PackageReference Include="Microsoft.AspNetCore.Components" Version="5.0.0" Condition= " '$(TargetFramework)' == 'net5.0' " />
	<PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="5.0.0" Condition= " '$(TargetFramework)' == 'net5.0' " />
</ItemGroup>
```

Again, not even the nuget package manager will correctly suggest the update 3.1.9 to 3.1.10, but rather to "consolidate" to 5.0.0.

### 3. Only targetting `net5.0`:

While this is the only situation where I can simply use the 5.X.Y packages, this is not a good idea if I want to not upset the users. Many won't update to net5.0 (e.g. when they use Mobile Blazor Bindings, or when their corporate policy is to stick with LTS releases), so targetting net5.0 only is not an option.

## Conclusion

Long story short: please consider supporting netstandard2.1 in the 5.X.Y packages.

**Comment by javiercn** [comment]
@stefanloerwald thanks for contacting us.

I believe it doesn't have to be that way. You could do something like:
```xml
<PropertyGroup>
	<TargetFrameworks>netstandard2.1;net5.0</TargetFrameworks>
        <RazorLangVersion Condition= " '$(TargetFramework)' != 'net5.0' ">3.0</RazorLangVersion>
        <ComponentsPackageVersion>3.1.9</ComponentsPackageVersion>
        <ComponentsPackageVersion Condition="'$(TargetFramework)' == 'net5.0' ">5.0.0</ComponentsPackageVersion>
</PropertyGroup>
<ItemGroup>
	<PackageReference Include="Microsoft.AspNetCore.Components" Version="$(ComponentsPackageVersion)" />
	<PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="$(ComponentsPackageVersion)" />
</ItemGroup>
```

Would that work?

> when their corporate policy is to stick with LTS releases

This only applies to users targeting Blazor server, Blazor Webassembly 3.2.0 is a "Current" release (not LTS) so those users need to move to 5.0.0

**Comment by stefanloerwald** [comment]
Hi @javiercn,

the issue is not about how to most elegantly write the package versions (that could also be done with two item groups). The issue is more with tooling, as it will suggest an update from 3.1.X to 5.0.0, even though that update will fail. And at the same time, it won't suggest the update from 3.1.9 to 3.1.10.

> > when their corporate policy is to stick with LTS releases
> 
> This only applies to users targeting Blazor server, Blazor Webassembly 3.2.0 is a "Current" release (not LTS) so those users need to move to 5.0.0

The case still stands: there will be people staying pre-net5.0, so dropping support for netstandard2.1 is not a good option.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @stefanloerwald.
There is no good solution here and we will not be able to do anything here due to the cost involved. Cross-compiling remains the only viable option for package authors, if they wish to publish packages targeting also .NET Standard.

**Comment by MagicAndre1981** [comment]
> ### 1. Staying on `netstandard2.1`:
> The nuget package manager and other tools (e.g. dependabot) will suggest the update to package version 5.0.0, even though that will fail due to incompatibility. There are now updates available to 3.1.X (namely currently 3.1.10), but they are less discoverable, because 5.0.0 is considered more recent by the package management tools.
> 

I can't help you with your request, but this is a [nuget issue](https://github.com/NuGet/Home/issues/6566) that Microsoft ignores for years. Old packages.config allowed you to limit version range, new PackageReference Include does not.

**Comment by stefanloerwald** [comment]
Agreed, @MagicAndre1981: this is mostly a nuget issue. It's just very much amplified by the incompatibility between netstandard2.1 and the 5.0.0 packages. It's even messing up dotnet format (https://github.com/dotnet/format/issues/862).

**Comment by MagicAndre1981** [comment]
I'm currently also doing the update of several nugets to 5.0 and test if it works in my old .net core 2.x/3.1 branches and hope I don't run into issues 🙏
